materialize-snowflake/snowflake.go

```diff
@@ -628,30 +660,42 @@ func (d *transactor) commit(ctx context.Context) error {
 		return err
 	}
 
+	group, groupCtx := errgroup.WithContext(ctx)
+	group.SetLimit(MaxConcurrentStores)
+
 	log.Info("store: starting copying of files into tables")
-	for _, b := range d.bindings {
-		if !b.store.stage.started {
-			// No table update required
-		} else if err := b.store.stage.flush(); err != nil {
-			return err
-		} else if !b.store.mustMerge {
-			log.WithField("table", b.target.Identifier).Info("store: starting direct copying data into table")
-			// We can issue a faster COPY INTO the target table.
-			if _, err = txn.ExecContext(ctx, b.store.copyInto); err != nil {
-				return fmt.Errorf("copying Store documents into table %q: %w", b.target.Identifier, err)
-			}
-			log.WithField("table", b.target.Identifier).Info("store: finishing direct copying data into table")
-		} else {
-			log.WithField("table", b.target.Identifier).Info("store: starting merging data into table")
-			// We must MERGE into the target table.
-			if _, err = txn.ExecContext(ctx, b.store.mergeInto); err != nil {
-				return fmt.Errorf("merging Store documents into table %q: %w", b.target.Identifier, err)
+	for _, bLoop := range d.bindings {
+		var b = bLoop
+		group.Go(func() error {
+			if !b.store.stage.started {
+				// No table update required
+			} else if err := b.store.stage.flush(); err != nil {
+				return err
+			} else if !b.store.mustMerge {
+				log.WithField("table", b.target.Identifier).Info("store: starting direct copying data into table")
+				// We can issue a faster COPY INTO the target table.
+				if _, err = txn.ExecContext(groupCtx, b.store.copyInto); err != nil {
```

Comment:
I have a similar concern here as the one above in `Load`: `txn` uses the same connection, and any concurrent operations with that connection will have to be run sequentially even if they are called from separate goroutines (assuming the `txn` is threadsafe - otherwise something even stranger might happen).

This is harder though, since we do need to run all of these operations in the same transaction to maintain exactly-once delivery.

TBH I'm not seeing how concurrent store operations are going to help us. One wild idea would be to switch to an idempotent apply style of storing, if that is possible with Snowflake, since that could probably be done concurrently. Or, we could concurrently load staged files into staging tables, and then in a single transaction merge the staging table data into the target tables. Both of these ideas are fraught with complexity and uncertainty of what if any benefit they would provide, so I'm certainly not advocating for them right now.

Reply:
That's a good point. I thought it is possible to run multiple statements in a single transaction concurrently so long as we committed after having run all of the statements. In that case I'll revert this part. I think the uploading of the files is most likely to have the most significant impact on performance improvement though.